src/session.c

```diff
@@ -689,44 +691,54 @@ CK_SLOT_ID p11prov_session_slotid(P11PROV_SESSION *session)
 
 /* returns a locked login_session if _session is not NULL */
 static CK_RV token_login(P11PROV_SESSION *session, P11PROV_URI *uri,
-                         OSSL_PASSPHRASE_CALLBACK *pw_cb, void *pw_cbarg)
+                         OSSL_PASSPHRASE_CALLBACK *pw_cb, void *pw_cbarg,
+                         struct p11prov_slot *slot)
 {
     char cb_pin[MAX_PIN_LENGTH + 1] = { 0 };
     size_t cb_pin_len = 0;
     CK_UTF8CHAR_PTR pin = NULL_PTR;
     CK_ULONG pinlen = 0;
     CK_RV ret;
 
-    if (uri) {
-        pin = (CK_UTF8CHAR_PTR)p11prov_uri_get_pin(uri);
-    }
-    if (!pin) {
-        pin = p11prov_ctx_pin(session->provctx);
-    }
-    if (pin) {
-        pinlen = strlen((const char *)pin);
-    } else if (pw_cb) {
-        const char *info = "PKCS#11 Token";
-        OSSL_PARAM params[2] = {
-            OSSL_PARAM_DEFN(OSSL_PASSPHRASE_PARAM_INFO, OSSL_PARAM_UTF8_STRING,
-                            (void *)info, sizeof(info)),
-            OSSL_PARAM_END,
-        };
-        ret = pw_cb(cb_pin, sizeof(cb_pin), &cb_pin_len, params, pw_cbarg);
-        if (ret != RET_OSSL_OK) {
-            ret = CKR_GENERAL_ERROR;
-            goto done;
+    if (!(slot->token.flags & CKF_PROTECTED_AUTHENTICATION_PATH)) {
+        if (uri) {
+            pin = (CK_UTF8CHAR_PTR)p11prov_uri_get_pin(uri);
+            if (slot->bad_pin && strcmp((char *)pin, slot->bad_pin) == 0) {
+                pin = NULL;
+            }
+        }
+        if (!pin) {
+            pin = p11prov_ctx_pin(session->provctx);
+            if (slot->bad_pin && strcmp((char *)pin, slot->bad_pin) == 0) {
+                pin = NULL;
+            }
         }
-        if (cb_pin_len == 0) {
+        if (pin) {
+            pinlen = strlen((const char *)pin);
+        } else if (pw_cb) {
+            const char *info = "PKCS#11 Token";
```

Comment:
I understand that this is just moved from existing code, but I think the prompt should have a bit more descriptive label. We are doing the login to the slot so slot label or something will be useful to distinguish which pin the user should provide.

Reply:
Have you tried an operation that requires prompting?
Just remove the pin from openssl.cnf and try one of the command that requires a pin and can get it from the user.

Reply:
Btw I think the PIN authentication is broken, or I forget which operations work with it (not all).

Reply:
Btw this is the prompt OpenSSL shows when it works (follow #172 for more fun):
> Enter pass phrase for PKCS#11 Token:

And now that I re-read your comment I understand you want to see something like:
"PKCS#11 Token (slot 1 - NSS Generic Crypto Services):" using the slot description field ?

Reply:
See #173 for this